### ISSUE TITLE
assertCount with remembered values causing false positives

### ISSUE THREAD

**Issue by janedbal** [phpstan/phpstan-phpunit#208]
Simplified reproducible example:

```php
class AssertCountProblemTest extends TestCase
{

    /**
     * @var list<string>
     */
    private static array $database = [];

    public function testFoo(): void
    {
        self::assertCount(0, $this->getDataFromDatabase());

        self::editDataInDatabase();

        $newData = $this->getDataFromDatabase();
        self::assertCount(1, $newData); // error: Call to static method PHPUnit\Framework\Assert::assertCount() with 1 and array{} will always evaluate to false.
    }

    /**
     * @return list<string>
     */
    private function getDataFromDatabase(): array
    {
        return self::$database;
    }

    private static function editDataInDatabase(): void
    {
        self::$database[] = 'new data';
    }

}
```

This poped up while upgrading phpstan-phpunit from `1.3.15` to `1.4.0`

**Comment by janedbal** [comment]
`1.3.16` is the version where this breaks.

**Comment by Ciloe** [comment]
Same here at version `1.4.1`.

I solved temporarly by creating a variable, and using it in the Count function : 

```php
// Before :
self::assertCount(1, $this->getDataFromDatabase());

// After :
$newData = $this->getDataFromDatabase();
self::assertCount(1, $newData);
```

**Comment by thePanz** [comment]
Just upgraded to PHPStan v2.11.1 and got this issue.
The same code worked fine with v2.10.7 (this was the previous version I had installed)

**Comment by ondrejmirtes** [comment]
@thePanz Your problem probably stems from a much smarter type narrowing when `count()` is involved. This isn't a bug but a feature. It's not specific to PHPUnit either.

You can either:

1) Mark the function as `@phpstan-impure`: https://phpstan.org/r/e621db72-ce3f-4833-a967-79c455f25542
2) or you can set `rememberPossiblyImpureFunctionValues: false`: https://phpstan.org/config-reference#rememberpossiblyimpurefunctionvalues

Detailed here: https://phpstan.org/blog/remembering-and-forgetting-returned-values

Also I talk about it for 10 minutes here: https://www.youtube.com/watch?v=AFjr3RlDOZQ&t=1728s

**Comment by ondrejmirtes** [comment]
As this isn't specific to PHPUnit either, @janedbal please re-report this in PHPStan: https://phpstan.org/r/b610c6cb-c572-49b1-a78c-1a1ad5bd9100

**Comment by janedbal** [comment]
> Please re-report this in PHPStan: https://phpstan.org/r/b610c6cb-c572-49b1-a78c-1a1ad5bd9100

Does that actually make sense? It works [when properly annotated](https://phpstan.org/r/7923a8c5-9f9b-46dd-acab-dcdf2fd55df3). I believe the problem is more about DX. Nobody wants to mark all similar methods with `@phpstan-impure`.

**Comment by ondrejmirtes** [comment]
I'm not sure what you're suggesting. As I'm describing the recording of my talk, these defaults make sense to me. If you don't want to use `@phpstan-impure`, there's `rememberPossiblyImpureFunctionValues: false`.
